include/kodlab_mjbots_sdk/pid_controller.h

```diff
@@ -0,0 +1,648 @@
+   * @param accumulator_limit Used to prevent runaway integrator/windup
+   * @param saturation controller output limits
+   */
+  PIDController(const VectorNS & p_gain,
+                const VectorNS & i_gain,
+                const VectorNS & d_gain, 
+                double time_step = 0.001,
+                VectorNS deadband = 0.05*VectorNS::Ones(), 
+                VectorNS accumulator_limit = 
+                    std::numeric_limits<Scalar>::max()*VectorNS::Ones(),
+                std::array<VectorNS, 2> saturation = 
+                    { -std::numeric_limits<Scalar>::max()*VectorNS::Ones(), 
+                       std::numeric_limits<Scalar>::max()*VectorNS::Ones() })
+    : 
+      deadband_(deadband),
+      accumulator_limit_(accumulator_limit),
+      saturation_(saturation),
+      time_step_(time_step){
+
+    set_gains(p_gain, i_gain, d_gain); 
+    error_ = VectorNS::Zero();
+    d_error_ = VectorNS::Zero();
+    set_setpoints(VectorNS::Zero(), VectorNS::Zero());
+  }
+  
+  /**
+   * @brief Construct a vector Pid object
+   * @param p_gain Proportional Gain
+   * @param i_gain Derivative Gain
+   * @param d_gain Integral Gain
+   * @param time_step Time difference between 2 updates
+   * @param deadband Used to set range within setpoints where no correction 
+   * input is applied
+   * @param accumulator_limit Used to prevent runaway integrator/windup
+   * @param saturation Used to set controller output limits
+   */
+  PIDController(Scalar p_gain,
+                Scalar i_gain,
+                Scalar d_gain, 
+                double time_step = 0.001,
+                VectorNS deadband = 0.05*VectorNS::Ones(),
+                VectorNS accumulator_limit = 
+                    std::numeric_limits<Scalar>::max()*VectorNS::Ones(),
+                std::array<VectorNS, 2> saturation = 
+                    { -std::numeric_limits<VectorNS>::max()*VectorNS::Ones(), 
+                       std::numeric_limits<VectorNS>::max()*VectorNS::Ones() })
+    : 
+      deadband_(deadband),
+      accumulator_limit_(accumulator_limit),
+      saturation_(saturation),
+      time_step_(time_step){
+
+    set_gains(p_gain, i_gain, d_gain); 
+    error_ = VectorNS::Zero();
+    d_error_ = VectorNS::Zero();
+    set_setpoints(VectorNS::Zero(), VectorNS::Zero());
+       
+  }
+
+  /**
+   * @brief Function to set the vector PID gains 
+   * @param p_gain Proportional Gain
+   * @param d_gain Derivative Gain
+   * @param i_gain Integral Gain
+   */
+  void set_gains(VectorNS p_gain, VectorNS i_gain, VectorNS d_gain) {
+    kp_ = p_gain;
+    ki_ = i_gain;
+    kd_ = d_gain;
+  }
+  /**
+   * @brief Function to set the scalar  PID gains
+   * @param p_gain Proportional Gain
+   * @param d_gain Derivative Gain
+   * @param i_gain Integral Gain
+   */
+  
+  void set_gains(Scalar p_gain, Scalar i_gain, Scalar d_gain) {
+    //checking if using for scalar PID
+    static_assert(N==1, "Using Scalar function in vector PID");
```

Comment:
You should make this assert a function since you will be using it in multiple places.

Reply:
Since this is a compile time check, the function may have requirements about how it is written. Look at asserts in Drake or Eigen or on google to check. My guess is a need for constexpr keyword .